File: julia/collatz.jl
#!/usr/bin/julia

function count_collatz(n)
    count = 0
    while n != 1
        if (n%2 == 1)
            n = 3*n + 1
        else
            n = n / 2
        end
        count = count + 1
    end 
    return count
end

# this program holds the number for which the collatz sequence length is to be computed,
# an Int128 called "num" and its corresponding sequence length, and Int in a tuple.
# Int128 is used for "num" as it's the largest integer data type in julia. 

collatz_list = Array{Tuple{Int128, Int}}(undef, 0)
# zero the list
for i = 1:10
    tup1 = (Int128(0), 0)
    push!(collatz_list, tup1)
end

# compute the collatz sequence lengths of each number up to 5 billion
num = Int128(1)                                     # number to compute collatz length of
max_value = Int128(10000)  
for i = 1:max_value
    global num
    tup = (num, count_collatz(num))
    if (tup[2] > collatz_list[10][2])               # determine if tup has a new max length to add
        offset = 1
        while (tup[2] < collatz_list[offset][2])    # determine where tup should go in array
            offset = offset + 1
        end

        temp = offset
        offset = 10
        while (offset > temp)                       # shift the other elements
            collatz_list[offset] = collatz_list[offset - 1]
            offset = offset - 1
        end
        collatz_list[temp] = tup                    # store tup in the array
    end
    num = num + 1
end

# print the collatz lengths sorted by length
println("The 10 largest collatz lengths sorted by length: ")
for i = 1:10
    println(collatz_list[i])
end

# sort the list by magnitude of num
sort!(collatz_list, by = x->x[1], rev = true)

# print the collatz lengths sorted by magnitude
println("The 10 largest collatz lengths sorted by magnitude: ")
for i = 1:10
    println(collatz_list[i])
end

File: ruby/collatz.rb
#!/usr/bin/ruby -w

Tuple = Struct.new(:num, :length)   # struct to hold integer/length pairs

def count_collatz(n)    # function to compute the collatz length for a given n
    count = 1
    while n != 1 do
        if n%2 == 1 then    # if odd
            n = 3*n + 1
        else                # if even
            n = n / 2
        end
        count = count + 1
    end
    return count
end 

def sort_length(array)
    n = array.length
    loop do
      swapped = false
  
      (n-1).times do |i|
        if array[i].length < array[i+1].length
          array[i], array[i+1] = array[i+1], array[i]
          swapped = true
        end
      end
  
      break if not swapped
    end
  
    array
end

def sort_magnitude(array)
    n = array.length
    loop do
      swapped = false
  
      (n-1).times do |i|
        if array[i].num < array[i+1].num
          array[i], array[i+1] = array[i+1], array[i]
          swapped = true
        end
      end
  
      break if not swapped
    end
  
    array
end

n = 1                       # holds value to be computed
collatz_list = Array.new    # array to store integer/length pairs

for i in 0..5000000000-1
    tup = Tuple.new
    tup.num = n
    tup.length = count_collatz(n)
    collatz_list[i] = tup
    n = n + 1
end

sort_length(collatz_list)

puts "The 10 longest collatz sequence lengths sorted by length: "
small = Array.new
for i in 0..9
    puts collatz_list[i]
    small[i] = collatz_list[i]
end

sort_magnitude(small)

puts "The 10 longest collatz sequence lengths sorted by magnitude: "
for i in 0..9
    puts small[i]
end